### ISSUE TITLE
fused_softmax kernel modifies input_grad tensor during backward step calculation.

### ISSUE THREAD

**Issue by yidong72** [NVIDIA/apex#1430]
**Describe the Bug**
In the current Apex fused softmax kernel implementation, I notice the backward pass modifies the `input_grad` tensor to be the same as the backward output tensor.
```C++
      dispatch_scaled_masked_softmax_backward<scalar_t, scalar_t, float>(
          reinterpret_cast<scalar_t*>(output_grads_ptr), 
	  reinterpret_cast<scalar_t*>(output_grads_ptr), 
	  reinterpret_cast<scalar_t const*>(softmax_results.data_ptr()),
	  scale_factor,
	  query_seq_len,
	  key_seq_len,
	  batches,
	  attn_heads);
```
However, according to the Pytorch document:
<img width="679" alt="image" src="https://user-images.githubusercontent.com/43824965/180485435-749e7745-c531-4f94-9504-e13a6114b539.png">
Pytorch document says explicitly that `It is important NEVER to modify these in-place`.  The current implementation break the contract which may cause some unpredicted problems. 
We should fix it and make it behave as the document says.
